tripy/tripy/frontend/trace/ops/pooling.py

```diff
@@ -0,0 +1,148 @@
+        padding = [(0, 0)] * extra_dims + list(self.padding)
+
+        ReduceWindowOp.build(
+            [inputs[0], init_const],
+            outputs,
+            reduce_mode=self.kind.op,
+            window_dims=window_dims,
+            window_strides=window_strides,
+            padding=padding,
+        )
+
+
+@export.public_api(document_under="operations/functions")
+@constraints.dtype_info(
+    dtype_variables={
+        "T1": ["float32", "float16", "int8"],
+    },
+    dtype_constraints={"input": "T1", constraints.RETURN_VALUE: "T1"},
+)
+def maxpool(
+    input: "tripy.Tensor",
+    kernel_dims: Sequence[int],
+    stride: Sequence[int] = None,
+    padding: Sequence[Tuple[int]] = None,
+) -> "tripy.Tensor":
+    r"""
+    Applies a max pooling over the input tensor.
+
+    The output's non-spatial dimensions are the same as input. For each input spatial dimension
+    :math:`D_{i}`, the corresponding output dimension will be:
+
+    .. math::
+        D_{out_i} = \left\lfloor\frac{D_{i} + \text{padding_before[i]} + \text{padding_after[i]} -
+                \text{kernel_dims[i]}}{\text{stride[i]}} + 1\right\rfloor
+
+    Args:
+        input: The input tensor.
+        kernel_dims: The spatial shape of the pooling window. Only 2-D or 3-D ``kernel_dims`` are supported.
+            If the input has ``int8`` datatype, ``kernel_dims`` can only be 2-D.
+        stride: A sequence of length :math:`M` indicating the stride of pooling across each spatial dimension,
+            where :math:`M` is the number of spatial dimensions, i.e. :math:`M = \text{rank(input)} - 2`.
+            Defaults to all 1.
+        padding: A sequence of pairs of integers of length :math:`M` indicating the zero padding
+            to apply to the input along each spatial dimension before and after the dimension respectively,
+            where :math:`M` is the number of spatial dimensions, i.e. :math:`M = \text{rank(input)} - 2`.
+            Defaults to all 0.
+
+    Returns:
+        The result tensor after the pooling operation.
+
+    .. code-block:: python
+        :linenos:
+        :caption: Example
+
+        input = tp.reshape(tp.arange(16, dtype=tp.float32), (1, 1, 4, 4))
+        output = tp.maxpool(input, kernel_dims=(2, 2))
+
+        pool_torch = torch.nn.MaxPool2d((2, 2), stride=1) # doc: omit
+        expected = pool_torch(torch.from_dlpack(input).to("cpu")) # doc: omit
+
+        assert torch.allclose(torch.from_dlpack(output).to("cpu"), expected)
+    """
+    from tripy.common.datatype import int8
+
+    spatial_dims = len(kernel_dims)
+    if spatial_dims != 2 and spatial_dims != 3:
+        raise_error("Unsupported kernel_dims, must be 2D or 3D.", [f"Got kernel_dims={kernel_dims}"])
+    if input.dtype == int8 and spatial_dims != 2:
+        raise_error(
+            "Unsupported kernel_dims for int8 dtype.",
+            [f"Only 2D kernel_dims is supported for int8, got {kernel_dims}."],
+        )
```

Comment:
Where does this restriction come from? I'd be more inclined to let it fail in the backend so that we're not artificially imposing a constraint.

Reply:
TRT does not support 3D max pooling for int8, it gives an error that I think is not that clean:
```
[optimizer.cpp::getFormatRequirements::3301] Error Code 2: Internal Error (Assertion !n->candidateRequirements.empty() failed. No supported formats for [tensorrt.pooling] (t15))
```
I can remove the frontend check if this is fine.